File: util/starutil.py
# This file is part of the Astrometry.net suite.
# Licensed under a 3-clause BSD style license - see LICENSE
from math import pi,cos,sin,radians,degrees,asin,atan2,sqrt,acos,floor
from functools import reduce

def hms2ra(h, m, s):
    return 15. * (h + (m + s/60.)/60.)

def dms2dec(sign, d, m, s):
    return sign * (d + (m + s/60.)/60.)

def tokenize_hms(s):
    s = s.strip()
    # split on whitespace
    tokens = s.split()
    # also split on colons
    tokens = reduce(list.__add__, [t.split(':') for t in tokens])
    assert(len(tokens) < 4)
    assert(len(tokens) > 0)
    h = len(tokens) >= 1 and float(tokens[0]) or 0
    m = len(tokens) >= 2 and float(tokens[1]) or 0
    s = len(tokens) >= 3 and float(tokens[2]) or 0
    return (h,m,s)

def hmsstring2hms(st):
    (h,m,s) = tokenize_hms(st)
    return (h,m,s)

def hmsstring2ra(st):
    return hms2ra(*hmsstring2hms(st))

# returns (sign, deg, minutes, second)
def dmsstring2dms(s):
    sign = 1.0
    if s[0] == '-':
        sign = -1.0
        s = s[1:]
    elif s[0] == '+':
        s = s[1:]
    (d,m,s) = tokenize_hms(s)
    return (sign, d, m, s)

def dmsstring2dec(s):
    return dms2dec(*dmsstring2dms(s))

# RA in degrees
def ra2hms(ra):
    h = ra * 24. / 360.
    hh = int(floor(h))
    m = (h - hh) * 60.
    mm = int(floor(m))
    s = (m - mm) * 60.
    return (hh, mm, s)

# Dec in degrees
# returns (sign, degrees, minutes, seconds)
def dec2dms(dec):
    sgn = (dec > 0) and 1 or -1
    d = dec * sgn
    dd = int(floor(d))
    m = (d - dd) * 60.
    mm = int(floor(m))
    s = (m - mm) * 60.
    return (sgn, d, m, s)

# RA in degrees
def ra2hmsstring(ra, separator=' '):
    (h,m,s) = ra2hms(ra)
    ss = int(floor(s))
    ds = int(round((s - ss) * 1000.0))
    return separator.join(['%0.2i' % h, '%0.2i' % m, '%0.2i.%0.3i' % (ss,ds)])

# Dec in degrees
def dec2dmsstring(dec, separator=' '):
    (sign,d,m,s) = dec2dms(dec)
    # whole number of seconds
    ss = int(floor(s))
    # milliseconds
    ds = int(round((s - ss) * 1000.0))
    # fix rounding errors...
    if ds >= 1000:
        ds -= 1000
        ss += 1
    return separator.join(['%+0.2i' % int(d*sign), '%0.2i' % m, '%0.2i.%0.3i' % (ss,ds)])

# RA, Dec in degrees
def radectoxyz(ra, dec):
    rarad = radians(ra)
    decrad = radians(dec)
    cosd = cos(decrad)
    return (cosd * cos(rarad), cosd * sin(rarad), sin(decrad))

# RA, Dec in degrees
def xyztoradec(x,y,z):
    return (degrees(xy2ra(x, y)), degrees(z2dec(z)))

def xyzarrtoradec(xyz):
    return (degrees(xy2ra(xyz[0], xyz[1])), degrees(z2dec(xyz[2])))

def rad2deg(r):    return 180.0*r/pi
def deg2rad(d):    return d*pi/180.0
def rad2arcmin(r): return 10800.0*r/pi
def arcmin2rad(a): return a*pi/10800.0
def rad2arcsec(r): return 648000.0*r/pi
def arcsec2rad(a): return a*pi/648000.0
def arcsec2deg(a): return rad2deg(arcsec2rad(a))
def radec2x(r,d):  return cos(d)*cos(r) # r,d in radians
def radec2y(r,d):  return cos(d)*sin(r) # r,d in radians
def radec2z(r,d):  return sin(d)        # r,d in radians
def z2dec(z):      return asin(z)     # result in radians
def xy2ra(x,y):
    "Convert x,y to ra in radians"
    r = atan2(y,x)
    r += 2*pi*(r<0.)
    return r

def degrees_between(ra1, dec1, ra2, dec2):
    return arcsec2deg(arcsec_between(ra1, dec1, ra2, dec2))

# RA,Decs in degrees.
def arcsec_between(ra1, dec1, ra2, dec2):
    xyz1 = radectoxyz(ra1, dec1)
    xyz2 = radectoxyz(ra2, dec2)
    d2 = sum([(a-b)**2 for (a,b) in zip(xyz1, xyz2)])
    return distsq2arcsec(d2)

def rad2distsq(rad):
    return 2. * (1. - cos(rad))

def arcsec2distsq(arcsec):
    return rad2distsq(arcsec2rad(arcsec))

def arcsec2dist(arcsec):
    return sqrt(arcsec2distsq(arcsec))

def dist2arcsec(dist):
    return distsq2arcsec(dist**2)

def distsq2arcsec(dist2):
    return rad2arcsec(distsq2rad(dist2))

def distsq2rad(dist2):
    return acos(1. - dist2 / 2.)

File: util/starutil.c
/*
 # This file is part of the Astrometry.net suite.
 # Licensed under a 3-clause BSD style license - see LICENSE
 */

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include <sys/types.h>
#include <regex.h>

#include "os-features.h"
#include "keywords.h"
#include "mathutil.h"
#include "starutil.h"
#include "errors.h"

#define POGSON 2.51188643150958
#define LOGP   0.92103403719762

#define InlineDefine InlineDefineC
#include "starutil.inc"
#undef InlineDefine

void radec_derivatives(double ra, double dec, double* dra, double* ddec) {
    double cosd = cos(deg2rad(dec));
    double cosra = cos(deg2rad(ra));
    double sinra = sin(deg2rad(ra));
    if (dra) {
        dra[0] = cosd * -sinra;
        dra[1] = cosd *  cosra;
        dra[2] = 0.0;
        normalize_3(dra);
    }
    if (ddec) {
        double sind = sin(deg2rad(dec));
        ddec[0] = -sind * cosra;
        ddec[1] = -sind * sinra;
        ddec[2] =  cosd;
        normalize_3(ddec);
    }
}

void radecrange2xyzrange(double ralo, double declo, double rahi, double dechi,
                         double* minxyz, double* maxxyz) {
    double minmult, maxmult;
    double uxlo, uxhi, uylo, uyhi;
    // Dec only affects z, and is monotonic (z = sin(dec))
    minxyz[2] = radec2z(0, declo);
    maxxyz[2] = radec2z(0, dechi);

    // min,max of cos(dec).  cos(dec) is concave down.
    minmult = MIN(cos(deg2rad(declo)), cos(deg2rad(dechi)));
    maxmult = MAX(cos(deg2rad(declo)), cos(deg2rad(dechi)));
    if (declo < 0 && dechi > 0)
        maxmult = 1.0;
    // unscaled x (ie, cos(ra))
    uxlo = MIN(cos(deg2rad(ralo)), cos(deg2rad(rahi)));
    if (ralo < 180 && rahi > 180)
        uxlo = -1.0;
    uxhi = MAX(cos(deg2rad(ralo)), cos(deg2rad(rahi)));
    minxyz[0] = MIN(uxlo * minmult, uxlo * maxmult);
    maxxyz[0] = MAX(uxhi * minmult, uxhi * maxmult);
    // unscaled y (ie, sin(ra))
    uylo = MIN(sin(deg2rad(ralo)), sin(deg2rad(rahi)));
    if (ralo < 270 && rahi > 270)
        uylo = -1.0;
    uyhi = MAX(sin(deg2rad(ralo)), sin(deg2rad(rahi)));
    if (ralo < 90 && rahi > 90)
        uyhi = -1.0;
    minxyz[1] = MIN(uylo * minmult, uylo * maxmult);
    maxxyz[1] = MAX(uyhi * minmult, uyhi * maxmult);
}


static int parse_hms_string(const char* str,
                            int* sign, int* term1, int* term2, double* term3) {
    anbool matched;
    regmatch_t matches[6];
    int nmatches = 6;
    regex_t re;
    regmatch_t* m;
    const char* s;

    const char* restr = 
        "^([+-])?([[:digit:]]{1,2}):"
        "([[:digit:]]{1,2}):"
        "([[:digit:]]*(\\.[[:digit:]]*)?)$";

    if (!str)
        return 1;

    if (regcomp(&re, restr, REG_EXTENDED)) {
        ERROR("Failed to compile H:M:S regex \"%s\"", restr);
        return -1;
    }
    matched = (regexec(&re, str, nmatches, matches, 0) == 0);
    regfree(&re);
    if (!matched)
        return 1;

    // sign
    m = matches + 1;
    s = str + m->rm_so;
    if (m->rm_so == -1 || s[0] == '+')
        *sign = 1;
    else
        *sign = -1;
    // hrs / deg
    m = matches + 2;
    s = str + m->rm_so;
    if (s[0] == '0')
        s++;
    *term1 = atoi(s);
    // Min
    m = matches + 3;
    s = str + m->rm_so;
    if (s[0] == '0')
        s++;
    *term2 = atoi(s);
    // Sec
    m = matches + 4;
    s = str + m->rm_so;
    *term3 = atof(s);
    return 0;
}

double atora(const char* str) {
    char* eptr;
    double ra;
    int sgn, hr, min;
    double sec;
    int rtn;

    if (!str) {
        //ERROR("Null string to atora()");
        return LARGE_VAL;
    }
    rtn = parse_hms_string(str, &sgn, &hr, &min, &sec);
    if (rtn == -1) {
        ERROR("Failed to run regex");
        return LARGE_VAL;
    }
    if (rtn == 0)
        return sgn * hms2ra(hr, min, sec);

    ra = strtod(str, &eptr);
    if (eptr == str)
        // no conversion
        return LARGE_VAL;
    return ra;
}

double atodec(const char* str) {
    char* eptr;
    double dec;
    int sgn, deg, min;
    double sec;
    int rtn;

    rtn = parse_hms_string(str, &sgn, &deg, &min, &sec);
    if (rtn == -1) {
        ERROR("Failed to run regex");
        return LARGE_VAL;
    }
    if (rtn == 0)
        return dms2dec(sgn, deg, min, sec);

    dec = strtod(str, &eptr);
    if (eptr == str)
        // no conversion
        return LARGE_VAL;
    return dec;
}

double mag2flux(double mag) {
    return pow(POGSON, -mag);
}

double flux2mag(double flux) {
    return -log(flux) * LOGP;
}

inline void xyzarr2radecarr(const double* xyz, double *radec) {
    xyz2radec(xyz[0], xyz[1], xyz[2], radec+0, radec+1);
}

double distsq_between_radecdeg(double ra1, double dec1,
                               double ra2, double dec2) {
    double xyz1[3];
    double xyz2[3];
    radecdeg2xyzarr(ra1, dec1, xyz1);
    radecdeg2xyzarr(ra2, dec2, xyz2);
    return distsq(xyz1, xyz2, 3);
}

double arcsec_between_radecdeg(double ra1, double dec1,
                               double ra2, double dec2) {
    return distsq2arcsec(distsq_between_radecdeg(ra1, dec1, ra2, dec2));
}

double deg_between_radecdeg(double ra1, double dec1, double ra2, double dec2) {
    return arcsec2deg(arcsec_between_radecdeg(ra1, dec1, ra2, dec2));
}

void project_equal_area(double x, double y, double z, double* projx, double* projy) {
    double Xp = x*sqrt(1./(1. + z));
    double Yp = y*sqrt(1./(1. + z));
    Xp = 0.5 * (1.0 + Xp);
    Yp = 0.5 * (1.0 + Yp);
    assert(Xp >= 0.0 && Xp <= 1.0);
    assert(Yp >= 0.0 && Yp <= 1.0);
    *projx = Xp;
    *projy = Yp;
}

void project_hammer_aitoff_x(double x, double y, double z, double* projx, double* projy) {
    double theta = atan(x/z);
    double r = sqrt(x*x+z*z);
    double zp, xp;
    /* Hammer-Aitoff projection with x-z plane compressed to purely +z side
     * of xz plane */
    if (z < 0) {
        if (x < 0) {
            theta = theta - M_PI;
        } else {
            theta = M_PI + theta;
        }
    }
    theta /= 2.0;
    zp = r*cos(theta);
    xp = r*sin(theta);
    assert(zp >= -0.01);
    project_equal_area(xp, y, zp, projx, projy);
}

/* makes a star object located uniformly at random within the limits given
 on the sphere */
void make_rand_star(double* star, double ramin, double ramax,
                    double decmin, double decmax)
{
    double decval, raval;
    if (ramin < 0.0)
        ramin = 0.0;
    if (ramax > (2*M_PI))
        ramax = 2 * M_PI;
    if (decmin < -M_PI / 2.0)
        decmin = -M_PI / 2.0;
    if (decmax > M_PI / 2.0)
        decmax = M_PI / 2.0;

    decval = asin(uniform_sample(sin(decmin), sin(decmax)));
    raval = uniform_sample(ramin, ramax);
    star[0] = radec2x(raval, decval);
    star[1] = radec2y(raval, decval);
    star[2] = radec2z(raval, decval);
}


// RA in degrees to Mercator X coordinate [0, 1).
inline double ra2mercx(double ra) {
    double mx = ra / 360.0;
    if (mx < 0.0 || mx > 1.0) {
        mx = fmod(mx, 1.0);
        if (mx < 0.0)
            mx += 1.0;
    }
    return mx;
}

// Dec in degrees to Mercator Y coordinate [0, 1).
inline double dec2mercy(double dec) {
    return 0.5 + (asinh(tan(deg2rad(dec))) / (2.0 * M_PI));
}

double hms2ra(int h, int m, double s) {
    return 15.0 * (h + ((m + (s / 60.0)) / 60.0));
}

double dms2dec(int sgn, int d, int m, double s) {
    return sgn * (d + ((m + (s / 60.0)) / 60.0));
}

// RA in degrees to H:M:S
inline void ra2hms(double ra, int* h, int* m, double* s) {
    double rem;
    ra = fmod(ra, 360.0);
    if (ra < 0.0)
        ra += 360.0;
    rem = ra / 15.0;
    *h = floor(rem);
    // remaining (fractional) hours
    rem -= *h;
    // -> minutes
    rem *= 60.0;
    *m = floor(rem);
    // remaining (fractional) minutes
    rem -= *m;
    // -> seconds
    rem *= 60.0;
    *s = rem;
}

// Dec in degrees to D:M:S
void dec2dms(double dec, int* sign, int* d, int* m, double* s) {
    double rem;
    double flr;
    *sign = (dec >= 0.0) ? 1 : -1;
    dec *= (*sign);
    flr = floor(dec);
    *d = flr;
    // remaining degrees:
    rem = dec - flr;
    // -> minutes
    rem *= 60.0;
    *m = floor(rem);
    // remaining (fractional) minutes
    rem -= *m;
    // -> seconds
    rem *= 60.0;
    *s = rem;
}

void ra2hmsstring(double ra, char* str) {
    int h, m;
    double s;
    int ss;
    int ds;
    ra2hms(ra, &h, &m, &s);

    // round to display to 3 decimal places
    ss = (int)floor(s);
    ds = (int)round((s - ss) * 1000.0);
    if (ds >= 1000) {
        ss++;
        ds -= 1000;
    }
    if (ss >= 60) {
        ss -= 60;
        m += 1;
    }
    if (m >= 60) {
        m -= 60;
        h += 1;
    }
    sprintf(str, "%02i:%02i:%02i.%03i", h, m, ss, ds);
}

void dec2dmsstring(double dec, char* str) {
    int sign, d, m;
    double s;
    int ss, ds;
    dec2dms(dec, &sign, &d, &m, &s);
    ss = (int)floor(s);
    ds = (int)round((s - ss) * 1000.0);
    if (ds >= 1000) {
        ss++;
        ds -= 1000;
    }
    if (ss >= 60) {
        ss -= 60;
        m += 1;
    }
    if (m >= 60) {
        m -= 60;
        d += 1;
    }
    sprintf(str, "%c%02i:%02i:%02i.%03i", (sign==1 ? '+':'-'), d, m, ss, ds);
}

